barretenberg/cpp/src/barretenberg/dsl/acir_format/avm2_recursion_constraint.test.cpp

```diff
@@ -0,0 +1,192 @@
+                }
+                return indices;
+            };
+
+            RecursionConstraint avm_recursion_constraint{
+                .key = add_to_witness_and_track_indices(key_witnesses),
+                .proof = add_to_witness_and_track_indices(proof_witnesses),
+                .public_inputs = add_to_witness_and_track_indices(public_inputs_witnesses),
+                .key_hash = 0, // not used
+                .proof_type = AVM,
+            };
+            avm_recursion_constraints.push_back(avm_recursion_constraint);
+        }
+
+        std::vector<size_t> avm_recursion_opcode_indices(avm_recursion_constraints.size());
+        std::iota(avm_recursion_opcode_indices.begin(), avm_recursion_opcode_indices.end(), 0);
+
+        AcirFormat& constraint_system = program.constraints;
+        constraint_system.varnum = static_cast<uint32_t>(witness.size());
+        constraint_system.num_acir_opcodes = static_cast<uint32_t>(avm_recursion_constraints.size());
+        constraint_system.avm_recursion_constraints = avm_recursion_constraints;
+        constraint_system.original_opcode_indices = create_empty_original_opcode_indices();
+
+        mock_opcode_indices(constraint_system);
+
+        return program;
+    }
+};
+
+TEST_F(AcirAvm2RecursionConstraint, TestBasicSingleAvm2RecursionConstraint)
+{
+    std::vector<InnerCircuitData> layer_1_circuits;
+    layer_1_circuits.push_back(create_inner_circuit_data());
+    AcirProgram avm_verifier_program = construct_avm_verifier_program(layer_1_circuits);
+    const ProgramMetadata metadata{ .honk_recursion = 1 };
+    auto layer_2_circuit = create_circuit(avm_verifier_program, metadata);
+
+    info("circuit gates = ", layer_2_circuit.get_estimated_num_finalized_gates());
+
+    auto proving_key = std::make_shared<DeciderProvingKey>(layer_2_circuit);
+    OuterProver prover(proving_key);
+    info("prover gates = ", proving_key->proving_key.circuit_size);
+    auto proof = prover.construct_proof();
+    auto verification_key = std::make_shared<OuterVerificationKey>(proving_key->proving_key);
+    OuterVerifier verifier(verification_key);
+    EXPECT_EQ(verifier.verify_proof(proof), true);
+}
+
+/**
+ * @brief Ensure that an AVM2 recursive verifier circuit VK can be constructed from a corresponding acir program without
+ * a witness.
+ * @details This is the logic required, for example, to write the VK of the public base circuit without knowledge of a
+ * particular satisfying witness.
+ *
+ */
+TEST_F(AcirAvm2RecursionConstraint, TestGenerateVKFromConstraintsWithoutWitness)
+{
+    // First, construct an AVM2 recursive verifier circuit VK by providing a valid program witness
+    std::shared_ptr<OuterVerificationKey> expected_vk;
+    {
+        auto layer_1_circuit = create_inner_circuit_data();
+        AcirProgram avm_verifier_program = construct_avm_verifier_program({ layer_1_circuit });
+        const ProgramMetadata metadata{ .honk_recursion = 1 };
```

Comment:
Just a comment for @ledwards2225: I know BB uses brace initialization a lot, but it's a bit of a can of worms (because it can mean many things). You'll find differing views on it on the internet, but in our codebase I'm trying to follow this: https://abseil.io/tips/88

Just an FYI. No need to change anything in this PR.